yt_dlp/downloader/youtube_live_chat.py

```diff
@@ -31,13 +32,116 @@ def real_download(self, filename, info_dict):
+                or try_get(item, lambda x: x['liveChatBannerRenderer'], dict)
+            )
+            if renderer is None:
+                return None
+            parent_item = (
+                try_get(
+                    renderer,
+                    lambda x: x['showItemEndpoint']['showLiveChatItemEndpoint']['renderer'], dict)
+                or try_get(renderer, lambda x: x['contents'], dict)
+            )
+            if parent_item:
+                renderer = (
+                    try_get(parent_item, lambda x: x['liveChatTextMessageRenderer'], dict)
+                    or try_get(parent_item, lambda x: x['liveChatPaidMessageRenderer'], dict)
+                    or try_get(parent_item, lambda x: x['liveChatMembershipItemRenderer'], dict)
+                    or try_get(parent_item, lambda x: x['liveChatPaidStickerRenderer'], dict)
+                )
+            if renderer:
+                timestamp_usec = try_get(renderer, lambda x: x['timestampUsec'], str)
+                if timestamp_usec:
+                    return int(int(timestamp_usec) / 1000)
+            return None
+
+        live_offset = [0]  # Python 2 doesnot support nonlocal
+
+        def parse_actions_live(live_chat_continuation):
+            continuation_id = None
+            processed_fragment = bytearray()
+            for action in live_chat_continuation.get('actions', []):
+                timestamp = parse_live_timestamp(action)
+                if timestamp is not None:
+                    live_offset[0] = timestamp - start_time
+                # compatibility with replay format
+                pseudo_action = {
+                    'replayChatItemAction': {'actions': [action]},
+                    'videoOffsetTimeMsec': str(live_offset[0]),
+                    'isLive': True,
+                }
+                processed_fragment.extend(
+                    json.dumps(pseudo_action, ensure_ascii=False).encode('utf-8') + b'\n')
+            continuation_data = (
+                try_get(
+                    live_chat_continuation,
+                    lambda x: x['continuations'][0]['invalidationContinuationData'], dict)
+                or try_get(
+                    live_chat_continuation,
+                    lambda x: x['continuations'][0]['timedContinuationData'], dict)
+            )
+            if continuation_data:
+                continuation_id = try_get(continuation_data, lambda x: x['continuation'])
+                timeout_ms = try_get(continuation_data, lambda x: x['timeoutMs'])
+                if timeout_ms is not None:
+                    time.sleep(int(timeout_ms) / 1000)
+            self._append_fragment(ctx, processed_fragment)
+            return continuation_id, live_offset[0]
+
+        if info_dict['protocol'] == 'youtube_live_chat_replay':
+            parse_actions = parse_actions_replay
+        elif info_dict['protocol'] == 'youtube_live_chat':
+            parse_actions = parse_actions_live
```

Comment:
Is it necessary to make 2 protocols? Can't you find this information from the `is_live` key?

Reply:
The `info_dict` here is not the complete one from `YoutubeIE._real_extract` but an item of `['subtitles']['live_chat']`. I don't know if the entire dict is available in this context. Alternatively `is_live` could be passed in as is and used for the same purpose

Reply:
oh, you are right. In that case, the current implementation is fine